File: src/tracker.py
import numpy as np
import torch
import cv2
from .feature.superpoint import SuperPoint
from .geometry.epipolar import EpipolarGeometry

class VisualOdometryTracker:
    """Core tracking system for Monocular Visual Odometry.

    Integrates the SuperPoint deep learning frontend for feature extraction
    and the epipolar geometry backend for camera pose estimation. Manages
    global trajectory states and keyframe selection based on parallax.
    """
    def __init__(self, config, weights_path):
        """Initializes the visual odometry tracker.

        Args:
            config (dict): Configuration dictionary containing model and camera parameters.
            weights_path (str): File path to the pre-trained SuperPoint weights.
        """
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        self.sp = SuperPoint(config['superpoint']).to(self.device)
        self.sp.load_state_dict(torch.load(weights_path, map_location=self.device, weights_only=True))
        self.sp.eval()
        
        self.geo = EpipolarGeometry(config['cam_intrinsics'])
        
        # Initialize global pose (Rotation and Translation matrices).
        self.cur_R = np.eye(3, dtype=np.float64)
        self.cur_t = np.zeros((3, 1), dtype=np.float64)
        
        # Store the state of the last registered keyframe.
        self.keyframe_kpts = None
        self.keyframe_desc = None
        
        # Parallax threshold for keyframe generation (in pixels).
        self.pixel_movement_threshold = 2.0

    def process_frame(self, frame_bgr):
        """Processes a single video frame to update the global camera trajectory.

        Extracts features, matches them against the last keyframe, evaluates
        parallax, and computes the relative pose if the baseline is sufficient.

        Args:
            frame_bgr (np.ndarray): The current BGR image frame from the video stream.

        Returns:
            tuple: A tuple containing:
                - current_translation (np.ndarray): The updated 3x1 global translation vector.
                - debug_image (np.ndarray): The frame annotated with tracked features.
        """
        gray = cv2.cvtColor(frame_bgr, cv2.COLOR_BGR2GRAY)
        img_tensor = torch.from_numpy(gray).float().unsqueeze(0).unsqueeze(0) / 255.0
        img_tensor = img_tensor.to(self.device)
        
        with torch.no_grad():
            out = self.sp(img_tensor)
            
        kpts = out['keypoints'][0].cpu().numpy()
        desc_map = out['descriptors'][0].cpu().numpy() 
        C, H_out, W_out = desc_map.shape
        
        kpts_scaled = kpts / 8.0
        kpts_scaled = np.clip(kpts_scaled, 0, [W_out-1, H_out-1]).astype(int)
        desc = desc_map[:, kpts_scaled[:, 1], kpts_scaled[:, 0]].T
        
        debug_img = frame_bgr.copy()
        for x, y in kpts:
            cv2.circle(debug_img, (int(x), int(y)), 3, (0, 255, 0), -1)

        # Register the first frame as the initial keyframe.
        if self.keyframe_kpts is None:
            self.keyframe_kpts = kpts
            self.keyframe_desc = desc
            return self.cur_t.copy(), debug_img

        # Match features between the current frame and the last keyframe.
        idx1, idx2 = self.geo.match_features(self.keyframe_desc, desc)
        
        if len(idx1) > 8:
            matched_kpts1 = self.keyframe_kpts[idx1]
            matched_kpts2 = kpts[idx2]
            
            # Calculate average optical flow magnitude (parallax) for tracked features.
            distances = np.linalg.norm(matched_kpts1 - matched_kpts2, axis=1)
            avg_dist = np.mean(distances)
            
            for pt1, pt2 in zip(matched_kpts1, matched_kpts2):
                cv2.line(debug_img, (int(pt1[0]), int(pt1[1])), (int(pt2[0]), int(pt2[1])), (0, 255, 255), 1)

            # Process geometry only if sufficient parallax (wide baseline) is observed.
            if avg_dist > self.pixel_movement_threshold:
                R, t, inlier_mask = self.geo.estimate_pose(matched_kpts1, matched_kpts2)
                
                if R is not None and t is not None:
                    # Inject pseudo-scale heuristic: map average pixel disparity to physical translation magnitude.
                    fake_scale = avg_dist / 20.0 
                    
                    # Update global pose using scaled translation and accumulated rotation.
                    self.cur_t = self.cur_t + self.cur_R.dot(t) * fake_scale
                    self.cur_R = self.cur_R.dot(R)
                    
                    # Update keyframe state.
                    self.keyframe_kpts = kpts
                    self.keyframe_desc = desc

        return self.cur_t.copy(), debug_img

File: src/tracker.cpp
#include "deepvo/tracker.h"
#include <numeric>
#include <algorithm>
#include <iostream>

namespace deepvo {
namespace tracker {

VisualOdometryTracker::VisualOdometryTracker(const cv::Mat& K)
    : geo_solver_(K), is_initialized_(false), K_(K.clone()), 
      current_scale_(1.0), prev_avg_dist_(-1.0) {
    
    // Initialize global pose at the origin facing forward
    cur_R_ = cv::Mat::eye(3, 3, CV_64F);
    cur_t_ = cv::Mat::zeros(3, 1, CV_64F);
}

void VisualOdometryTracker::DetectNewFeatures(const cv::Mat& frame_gray, std::vector<cv::Point2f>& kpts) {
    // Utilize Shi-Tomasi corner detection, which provides excellent stability for optical flow
    cv::goodFeaturesToTrack(
        frame_gray, kpts, 
        kMaxFeatures, 
        0.01,                 // Quality level
        kMinFeatureDistance,  // Enforce uniform spatial distribution
        cv::Mat(), 
        3,                    // Block size
        false, 
        0.04
    );
}

bool VisualOdometryTracker::ProcessFrame(const cv::Mat& frame) {
    cv::Mat frame_gray;
    if (frame.channels() == 3) {
        cv::cvtColor(frame, frame_gray, cv::COLOR_BGR2GRAY);
    } else {
        frame_gray = frame.clone();
    }

    // ---------------------------------------------------------
    // Phase 1: System Initialization
    // ---------------------------------------------------------
    if (!is_initialized_) {
        DetectNewFeatures(frame_gray, prev_kpts_);
        prev_frame_gray_ = frame_gray;
        is_initialized_ = true;
        std::cout << "[Tracker] System initialized with " << prev_kpts_.size() << " features." << std::endl;
        return true;
    }

    // ---------------------------------------------------------
    // Phase 2: KLT Optical Flow Tracking
    // ---------------------------------------------------------
    std::vector<cv::Point2f> cur_kpts;
    std::vector<uchar> status;
    std::vector<float> err;

    // Calculate sparse optical flow utilizing image pyramids
    cv::calcOpticalFlowPyrLK(
        prev_frame_gray_, frame_gray, 
        prev_kpts_, cur_kpts, 
        status, err, 
        cv::Size(21, 21), 
        3 // Max pyramid level
    );

    // Filter out points that the KLT algorithm failed to track
    std::vector<cv::Point2f> matched_prev, matched_cur;
    double total_dist = 0.0;
    
    for (size_t i = 0; i < status.size(); i++) {
        if (status[i]) {
            matched_prev.push_back(prev_kpts_[i]);
            matched_cur.push_back(cur_kpts[i]);
            
            // Accumulate pixel displacement to estimate relative physical scale later
            total_dist += cv::norm(cur_kpts[i] - prev_kpts_[i]);
        }
    }

    if (matched_cur.empty()) {
        std::cerr << "[Tracker] Fatal: Complete loss of tracking." << std::endl;
        is_initialized_ = false;
        return false;
    }

    double avg_dist = total_dist / matched_cur.size();

    // ---------------------------------------------------------
    // Phase 3: Motion Estimation and Implicit Keyframe Management
    // ---------------------------------------------------------
    
    // Require a minimum average pixel displacement to trigger pose estimation.
    // This threshold filters out micro-jitter and ensures a sufficiently wide 
    // baseline for accurate epipolar geometry computation.
    if (avg_dist > 2.0) {
        cv::Mat R, t, inlier_mask;
        std::tie(R, t, inlier_mask) = geo_solver_.EstimatePose(matched_prev, matched_cur);

        // Ensure the pose recovery succeeded and is supported by a robust number of inliers.
        // This prevents aggressive trajectory jumps caused by severe motion blur or dynamic objects.
        if (!R.empty() && !t.empty() && cv::countNonZero(inlier_mask) > 15) {
            
            // Inject pseudo-scale heuristic: map average pixel disparity to physical translation magnitude.
            // This dynamic scaling allows the monocular system to approximate velocity changes.
            double fake_scale = avg_dist / 20.0; 

            // Convert matrices to double precision (CV_64F) prior to multiplication 
            // to maintain strict numerical stability.
            R.convertTo(R, CV_64F);
            t.convertTo(t, CV_64F);

            // Update the global trajectory.
            // Strict rotation limits are deliberately omitted here to accommodate 
            // aggressive cornering movements in the dataset.
            cur_t_ = cur_t_ + fake_scale * (cur_R_ * t);
            cur_R_ = cur_R_ * R;

            // Implicit Keyframe Mechanism:
            // Only update the reference features (prev_kpts_) when a valid camera movement is recorded.
            // If the camera is nearly stationary, the KLT tracker will safely continue tracking 
            // from the last valid keyframe, accumulating disparity until it crosses the threshold.
            prev_kpts_ = matched_cur;
            prev_frame_gray_ = frame_gray;
        }
    }

    // ---------------------------------------------------------
    // Phase 4: Feature Pool Management
    // ---------------------------------------------------------
    
    // Replenish the feature pool if a significant number of points are lost 
    // due to occlusion, severe rotation, or objects exiting the field of view.
    if (prev_kpts_.size() < kMinFeatures) {
        DetectNewFeatures(frame_gray, prev_kpts_);
        prev_frame_gray_ = frame_gray;
        std::cout << "[Tracker] Feature pool replenished. Current size: " << prev_kpts_.size() << std::endl;
    }

    return true;
}

cv::Mat VisualOdometryTracker::GetCurrentR() const { return cur_R_.clone(); }
cv::Mat VisualOdometryTracker::GetCurrentT() const { return cur_t_.clone(); }
std::vector<cv::Point2f> VisualOdometryTracker::GetTrackedPoints() const { return prev_kpts_; }

}  // namespace tracker
}  // namespace deepvo